src/raster/r.viewshed.exposure/r.viewshed.exposure.py

```diff
@@ -0,0 +1,1409 @@
+    :return: 2D array of weighted parametrised viewshed
+    :rtype: ndarray
+    """
+    # 1. Convert DSM to numpy
+    np_dsm = raster2numpy(r_dsm)
+
+    # Ensure that values are represented as float (in case of CELL
+    # data type) and replace integer NaN with numpy NaN
+    dsm_type = grass.parse_command("r.info", map=r_dsm, flags="g")["datatype"]
+
+    if dsm_type == "CELL":
+        np_dsm = np_dsm.astype(np.float32)
+        np_dsm[np_dsm == -2147483648] = np.nan
+
+    # 2. local row, col coordinates and global Z coordinate of observer points V
+    #    3D array (lreg_shape[0] x lreg_shape[1] x 3)
+    v_loc = np.array(
+        [
+            np.tile(
+                np.arange(0.5, lreg_shape[0] + 0.5).reshape(-1, 1), (1, lreg_shape[1])
+            ),
+            np.tile(
+                np.arange(0.5, lreg_shape[1] + 0.5).reshape(-1, 1).transpose(),
+                (lreg_shape[0], 1),
+            ),
+            np_dsm + v_elevation,
+        ]
+    )
+
+    # 3. vector VT, adjusted for cell size
+    #    3D array (lreg_shape[0] x lreg_shape[1] x 3)
+    v_vect = np.array(
+        [
+            (v_loc[0] - t_loc[0]) * reg.nsres,
+            (v_loc[1] - t_loc[1]) * reg.ewres,
+            (v_loc[2] - t_loc[2]),
+        ]
+    )
+
+    # 4. projection of vector VT to XZ and YZ plane, adjusted for cell size
+    v_vect_ns = np.array([(v_loc[0] - t_loc[0]) * reg.nsres, v_loc[2] - t_loc[2]])
+
+    v_vect_ew = np.array([(v_loc[1] - t_loc[1]) * reg.ewres, v_loc[2] - t_loc[2]])
+
+    v_vect_ns_unit = v_vect_ns / np.linalg.norm(v_vect_ns, axis=0)
+    v_vect_ew_unit = v_vect_ew / np.linalg.norm(v_vect_ew, axis=0)
+
+    # 5. size of vector VT
+    #    2D array (lreg_shape[0] x lreg_shape[1])
+    v_scal = np.sqrt(v_vect[0] ** 2 + v_vect[1] ** 2 + v_vect[2] ** 2)
+    # replace 0 distance for central pixel by resolution to avoid division by 0
+    # v_scal = np.where(abs(v_scal) < 1e-6, (reg.nsres + reg.ewres) / 2, v_scal)
+    # grass.verbose(v_scal)
+
+    # 6. vector n, its projection to XZ, YZ plane
+    #   1D array [X, Z], [Y, Z]
+    #   already unit vector
+    n_vect_ns_unit = [np.sin(np.radians(t_loc[4])), np.cos(np.radians(t_loc[4]))]
+    n_vect_ew_unit = [np.sin(np.radians(t_loc[3])), np.cos(np.radians(t_loc[3]))]
+
+    # 7. angles beta (ns), theta (ew) (0-90 degrees)
+    #    2D array (lreg_shape[0] x lreg_shape[1])
+    beta = np.arccos(
+        n_vect_ns_unit[0] * v_vect_ns_unit[:][0]
+        + n_vect_ns_unit[1] * v_vect_ns_unit[:][1]
+    )
+    beta = np.where(beta > math.pi / 2, beta - math.pi / 2, beta)
+
+    theta = np.arccos(
+        n_vect_ew_unit[0] * v_vect_ew_unit[:][0]
+        + n_vect_ew_unit[1] * v_vect_ew_unit[:][1]
+    )
+    theta = np.where(theta > math.pi / 2, theta - math.pi / 2, theta)
+
+    # 8. visual magnitude adjusted for distance weight
+    visual_magnitude = (
+        np.cos(beta) * np.cos(theta) * ((reg.nsres * reg.ewres) / (v_scal ** 2))
+    )
+    sys.stderr.flush()
```

Comment:
Why is this needed here?

Reply:
I think it is a remnant from earlier progress tracking (right, @zofie-cimburova )... Removed it now in e45e875. We will see if it was needed...